### PR TITLE
Removed protection hack and replaced with logic from PAL dump.

### PR DESCRIPTION
**Pull request by va7deo**
The protection workaround was replaced with emulation of the PAL16R6.  The logic equations were based on the PAL dump and output of jedutil.  Each write to the tile ram latches a new value on the output of the U001 protection IC which is read by the CPU during the protection check.

### CODE REVIEW

File: src/mame/video/mrdo.cpp [hunk 1]
```diff
@@ -185,13 +185,61 @@ void mrdo_state::mrdo_bgvideoram_w(offs_t offset, uint8_t data)
 	m_bg_tilemap->mark_tile_dirty(offset & 0x3ff);
 }
 
+uint8_t pal_u001;
```
Comment (0kmg) [inline src/mame/video/mrdo.cpp:188]: This needs to be encapsulated within the base class, not visible at global scope. In that case note the MAME convention of prefixing class variables, i.e. `m_pal_u001`.

Comment (va7deo) [reply src/mame/video/mrdo.cpp]: I moved it to class scope and renamed.

File: src/mame/video/mrdo.cpp [hunk 2]
```diff
@@ -185,13 +185,59 @@ void mrdo_state::mrdo_bgvideoram_w(offs_t offset, uint8_t data)
...
+    return m_pal_u001;
+}
+
 void mrdo_state::mrdo_fgvideoram_w(offs_t offset, uint8_t data)
 {
 	m_fgvideoram[offset] = data;
 	m_fg_tilemap->mark_tile_dirty(offset & 0x3ff);
+    
+    // protection.  each write latches a new value on IC u001 (PAL16R6)
+    uint8_t i9 = data & 0x01;
```
Comment (jackson2k2) [inline src/mame/video/mrdo.cpp:201]: These statements should be done with the BIT template expressions we have; i.e. **BIT(data, 0)** and so on.

Comment (va7deo) [reply src/mame/video/mrdo.cpp]: It has been updated.  I looked for contribution docs but it looks like the page is broken. 
https://wiki.mamedev.org/index.php/Submitting_Source_Code  
The docs/CONTRIBUTING.md is also empty.  Are there other resources to help with code submissions?

Comment (Tafoid) [reply src/mame/video/mrdo.cpp]: Sadly, that page is still malfunctioning.  If you have code or data to send, use
Submit-email

File: src/mame/video/mrdo.cpp (hunk 1 above)
Comment (DavidHaywood) [reply src/mame/video/mrdo.cpp]: it will also need registering to be saved in the machine_start, and initialized, probably in machine_reset

Comment (va7deo) [reply src/mame/video/mrdo.cpp]: Added m_pal_u001 save state handling.

File: src/mame/video/mrdo.cpp [hunk 3]
```diff
@@ -185,13 +185,59 @@ void mrdo_state::mrdo_bgvideoram_w(offs_t offset, uint8_t data)
...
+    uint8_t r12 = 0;
+    uint8_t r13 = (1^( t1 ))      << 1 ;
+    uint8_t r14 = (1^( t1 | t2 )) << 2 ;
+    uint8_t r15 = (1^( t1 | t3 )) << 3 ;
+    uint8_t r16 = (1^( t1 ))      << 4 ;
+    uint8_t r17 = (1^( t1 | t3 )) << 5 ;
+    uint8_t r18 = (1^( t3 | t4 )) << 6 ;
+    uint8_t r19 = 0;
+    
+    m_pal_u001 = r19 | r18 | r17 | r16 | r15 | r14 | r13 | r12  ;
```
Comment (cuavas) [inline src/mame/video/mrdo.cpp:225]: Please use a tab for the initial indent.  Also, you can `const`-qualify all your intermediate values to make it obvious they’re not supposed to be modified after they’re set.

**Review by cuavas — CHANGES_REQUESTED** [review]

File: src/mame/includes/mrdo.h [hunk 4]
```diff
@@ -47,8 +50,10 @@ class mrdo_state : public driver_device
 	tilemap_t *m_bg_tilemap;
 	tilemap_t *m_fg_tilemap;
 	int       m_flipscreen;
-
-	uint8_t mrdo_SECRE_r();
+    uint8_t   m_pal_u001;
+    
+    uint8_t mrdo_SECRE_r();
+
```
[the remark this answers is not in the document]
Comment (va7deo) [reply src/mame/includes/mrdo.h]: Changes have been made.

**Comment by p-v-s** [comment]
Hello.  I just thought I would mention that MRDO no longer works properly as of MAME 0.240.  Once the player gets an extra man, only one of the "EXTRA" letters are reset. In addition, the advance of the "goblins" when the EXTRA guy is loose are messed up by apples and other things once one EXTRA man has been earned.

I have double checked all of my downloads (mame binary and romset), and compared the authentic binary behavior to that of a personally-compiled linux version (of 0.240).  MRDO behaves incorrectly in both versions of MAME0.240, but was fine up through 0.239.

Sorry if this is not the correct place to post this.  I joined github just to let you know about this issue, since I cannot find any mention of it elsewhere.

**Comment by p-v-s** [comment]
> This should already be fixed, see commit [8cc6459](https://github.com/mamedev/mame/commit/8cc6459f4c52f2620d813607027fd7275194fe4a).

What can I say?!  WOW!!!  You "guys" are absolutely AMAZING!  Thank you so much.  I will try to (again, new to GitHub) DL the affected the files and build it into my Linux compilation.  At worst ... wait until 0.241. ;-)

Thanks again!!!
-pvs

**Pull request merged**